File: src/main/java/com/tim/sernachat/repository/ChatRoomRepository.java
package com.tim.sernachat.repository;

import com.tim.sernachat.model.ChatRoom;
import lombok.RequiredArgsConstructor;
import org.springframework.data.redis.core.HashOperations;
import org.springframework.data.redis.core.ValueOperations;
import org.springframework.stereotype.Repository;

import java.util.List;
import java.util.Optional;
import javax.annotation.Resource;

@RequiredArgsConstructor
@Repository
public class ChatRoomRepository {
    public static final String USER_COUNT = "USER_COUNT";
    public static final String ENTER_INFO = "ENTER_INFO";
    // Redis CacheKeys
    private static final String CHAT_ROOMS = "CHAT_ROOM";
    @Resource(name = "redisTemplate")
    private HashOperations<String, String, ChatRoom> hashOpsChatRoom;
    @Resource(name = "redisTemplate")
    private HashOperations<String, String, String> hashOpsEnterInfo;
    @Resource(name = "redisTemplate")
    private ValueOperations<String, String> valueOps;

    public ChatRoom createChatRoom(String name) {
        ChatRoom chatRoom = ChatRoom.create(name);
        hashOpsChatRoom.put(CHAT_ROOMS, chatRoom.getRoomId(), chatRoom);
        return chatRoom;
    }

    public List<ChatRoom> findAllRoom() {
        return hashOpsChatRoom.values(CHAT_ROOMS);
    }

    public ChatRoom findRoomById(String id) {
        return hashOpsChatRoom.get(CHAT_ROOMS, id);
    }

    public void setUserEnterInfo(String sessionId, String roomId) {
        hashOpsEnterInfo.put(ENTER_INFO, sessionId, roomId);
    }

    public String getUserEnterRoomId(String sessionId) {
        return hashOpsEnterInfo.get(ENTER_INFO, sessionId);
    }

    public void removeUserEnterInfo(String sessionId) {
        hashOpsEnterInfo.delete(ENTER_INFO, sessionId);
    }

    public long getUserCount(String roomId) {
        return Long.valueOf(Optional.ofNullable(valueOps.get(USER_COUNT + "_" + roomId)).orElse("0"));
    }

    public long plusUserCount(String roomId) {
        return Optional.ofNullable(valueOps.increment(USER_COUNT + "_" + roomId)).orElse(0L);
    }

    public long minusUserCount(String roomId) {
        return Optional.ofNullable(valueOps.decrement(USER_COUNT + "_" + roomId)).filter(count -> count > 0).orElse(0L);
    }
}

File: src/main/kotlin/com/itime/sernachat/repository/ChatRoomRepository.kt
package com.itime.sernachat.repository

import org.springframework.data.redis.core.HashOperations
import org.springframework.data.redis.core.ValueOperations
import org.springframework.stereotype.Repository
import java.util.*
import javax.annotation.Resource

//@Repository
//class ChatRoomRepository {
//    @Resource(name = "redisTemplate")
//    private val hashOpsChatRoom: HashOperations<String, String, ChatRoom>? = null
//
//    @Resource(name = "redisTemplate")
//    private val hashOpsEnterInfo: HashOperations<String, String, String>? = null
//
//    @Resource(name = "redisTemplate")
//    private val valueOps: ValueOperations<String, String>? = null
//    fun createChatRoom(name: String?): ChatRoom {
//        val chatRoom: ChatRoom = ChatRoom.create(name)
//        chatRoom.roomId?.let { hashOpsChatRoom!!.put(CHAT_ROOMS, it, chatRoom) }
//        return chatRoom
//    }
//
//    fun findAllRoom(): List<ChatRoom> {
//        return hashOpsChatRoom!!.values(CHAT_ROOMS)
//    }
//
//    fun findRoomById(id: String?): ChatRoom? {
//        return hashOpsChatRoom!![CHAT_ROOMS, id!!]
//    }
//
//    fun setUserEnterInfo(sessionId: String, roomId: String) {
//        hashOpsEnterInfo!!.put(ENTER_INFO, sessionId, roomId)
//    }
//
//    fun getUserEnterRoomId(sessionId: String?): String? {
//        return hashOpsEnterInfo!![ENTER_INFO, sessionId!!]
//    }
//
//    fun removeUserEnterInfo(sessionId: String?) {
//        hashOpsEnterInfo!!.delete(ENTER_INFO, sessionId)
//    }
//
//    fun getUserCount(roomId: String): Long {
//        return java.lang.Long.valueOf(Optional.ofNullable(valueOps!![USER_COUNT + "_" + roomId]).orElse("0"))
//    }
//
//    fun plusUserCount(roomId: String): Long {
//        return Optional.ofNullable(valueOps!!.increment(USER_COUNT + "_" + roomId)).orElse(0L)
//    }
//
//    fun minusUserCount(roomId: String): Long {
//        return Optional.ofNullable(valueOps!!.decrement(USER_COUNT + "_" + roomId)).filter { count: Long -> count > 0 }
//            .orElse(0L)
//    }
//
//    companion object {
//        const val USER_COUNT = "USER_COUNT"
//        const val ENTER_INFO = "ENTER_INFO"
//
//        // Redis CacheKeys
//        private const val CHAT_ROOMS = "CHAT_ROOM"
//    }
//}